Fixed compilation with msvc

arves100 opened the pull request:
Under Microsoft Visual C++ compiler, the arguments that starts with - does not exists. (They are exclusive to GCC)
There was also a mistake in include_directories for GLEW.

-O3 is not necessary (counterpart is /Ox) since CMake can Debug/Release builds can be switched inside Vissual Studio
-Wno-errors is disabled by default (/WX-)
narrowing flags are enabled with /W4, default one in CMake is /W3
funroll loops should be enabled when you do a Release build (with optimization)

EDIT: The second commit fixes the compilation issue I've run into visual studio.
- Wrapped "int main" into SDL_main (used by library SDL2_Main) because MSVC may get confused without having WinMain, SDL2 allows to keep compatibility with WinMain (windows specific entry point) and other application by using SDL_main. (I remember that could also be fixed by specifing /SUBSYSTEM:CONSOLE in the linker properties, but it seems like it's not working right)
- Removed Variable Array Length, they are not supported in Visual C++.

shonumi commented on src/dmg/apu.cpp:
```diff
@@ -584,10 +584,10 @@ void dmg_audio_callback(void* _apu, u8 *_stream, int _length)
+	s16* channel_1_stream = new s16[length];
+	s16* channel_2_stream = new s16[length];
+	s16* channel_3_stream = new s16[length];
+	s16* channel_4_stream = new s16[length];
```
This part is kinda ugly. I guess MSVC doesn't like switching to C99... Anyway, something like this is preferred:

    std::vector<s16> channel_1_stream(length);
    std::vector<s16> channel_2_stream(length);
    etc...

This avoids manual memory management and gets rid of having to call **delete**

shonumi commented on src/dmg/apu.cpp:
```diff
@@ -643,4 +643,9 @@ void dmg_audio_callback(void* _apu, u8 *_stream, int _length)
+	delete channel_4_stream;
+	delete channel_3_stream;
+	delete channel_2_stream;
+	delete channel_1_stream;
```
With std::vector, these aren't necessary. See above comment.

shonumi commented on src/gba/apu.cpp:
```diff
@@ -365,12 +365,13 @@ void agb_audio_callback(void* _apu, u8 *_stream, int _length)
-	s16 channel_3_stream[length];
-	s16 channel_4_stream[length];
-	s16 dma_a_stream[length];
-	s16 dma_b_stream[length];
```
Same as above comments. If MSVC doesn't support Variable Length Arrays, C++ vectors are preferred to using **new** and **delete**.

shonumi commented on src/sgb/core.cpp:
```diff
@@ -935,13 +935,13 @@ void SGB_core::dump_obj(int obj_index) { }
 
 /****** Grabs the OBJ palette ******/
-u32* SGB_core::get_obj_palette(int pal_index) { }
+u32* SGB_core::get_obj_palette(int pal_index) { return nullptr; }
```
Just NULL for now, until GBE+ moves to C++11 or newer. I don't know when/if that'll ever happen.

shonumi commented on src/sgb/core.cpp:
```diff
@@ -935,13 +935,13 @@ void SGB_core::dump_obj(int obj_index) { }
 
 /****** Grabs the BG palette ******/
-u32* SGB_core::get_bg_palette(int pal_index) { }
+u32* SGB_core::get_bg_palette(int pal_index) { return nullptr; }
```
Same as previous comment.

shonumi commented:
Thanks for the contribution!

I just have a few things to point out:

1) I'd rather not use **new** and **delete** for arrays. I try to avoid manually managing memory. C++ vectors are preferred. Variable Length Arrays (VLAs) are supported by gcc/g++, but not MSVC apparently. C++ vectors works best for this situation.

2) Using NULL is preferable for now instead of nullptr. If and when GBE+ changes over to C++11 or newer, I'll change them all at once. Might seem kinda silly to others, but I had a specific plan to make C++11 changes, and I'd like to make sure I don't mix and match NULL and nullptr until that time.

3) When you make the above changes, be sure to [squash your multiple commits into a single commit](https://coderwall.com/p/5_cjhw/squash-your-commits-with-git-rebase)

shonumi commented on src/main.cpp:
```diff
@@ -14,7 +14,9 @@
-int main(int argc, char* args[]) 
+#include <SDL2/SDL_main.h>
+
+int main(int argc, char** args)
```
Try to keep this as:
    int main(int argc, char *argv[])

arves100 commented:
- Changed VLA to std::vectors
- Fixed linking issues with gbe_plus_qt
- Forced /SUBSYSTEM:CONSOLE when generating exe
- Changed nullptr to NULL.

Note: I didn't use std::vector.data() function since you clarified that gbe+ doesn't use C++11.
Second note: I've not added any resource file, Windows uses the default icon for GBE+ (when you're trying to open the application, they have the default icon), but this does not affect the task bar icon (since it's setted up by SDL2 or Qt)

shonumi commented:
Looks good! Let me check some things when I get back to my main computer in a few hours, then I'll merge it. Good work!
